crates/primitives/fee/src/lib.rs

```diff
@@ -81,21 +81,27 @@ pub fn compute_transaction_resources<S: State + StateChanges>(
 }
 
 /// Charges the fees for a specific execution resources.
-pub fn charge_fee<S: State + StateChanges + FeeConfig>(
+pub fn charge_fee<S: State + StateChanges + StateConfigProvider>(
     state: &mut S,
     block_context: &BlockContext,
     account_tx_context: AccountTransactionContext,
     resources: &ResourcesMapping,
+    disable_fee_charge: bool,
 ) -> TransactionExecutionResult<(Fee, Option<CallInfo>)> {
+    // disable_transaction_fee in the config level implies that transaction fees have been disabled
+    // at the app level and so we return 0 as the fees
     if state.is_transaction_fee_disabled() {
         return Ok((Fee(0), None));
     }
 
     let actual_fee = calculate_tx_fee(resources, block_context)?;
 
-    // even if the user doesn't have enough balance
-    // estimate fee shouldn't fail
-    if account_tx_context.version.0 >= StarkFelt::try_from("0x100000000000000000000000000000000").unwrap() {
+    // if the tx version is that of a estimate tx, then don't charge any fees
+    // if `disable_fee_charge` is true, it means its a simulate transaction so
+    // we don't charge fees but still return the correct fee amount
```

Comment:
You can say it better. It's a bit messy here. I struggle to understand what you mean